### DISCUSSION TITLE
Unable to resolve type Avalonia.Data.RelativeSource

### DISCUSSION THREAD

**Discussion by denver116** [AvaloniaUI/Avalonia#14992] [Q&A] [answered]
Hi, I'm new to Avalonia and trying to convert a xaml file to an image. Here is my program

                var xaml = File.ReadAllText(@"INPUT.xaml");

                var visual = new UserControl();
                using (var stream = new MemoryStream())
                using (var writer = new StreamWriter(stream))
                {
                    writer.Write(xaml);
                    writer.Flush();
                    stream.Position = 0;

                    visual = (UserControl)AvaloniaRuntimeXamlLoader.Load(stream);
                }

                visual.Measure(Size.Infinity);
                visual.Arrange(new Rect(visual.DesiredSize));

                var bounds = visual.Bounds;
                var dpi = new Vector(400, 400);
                var pixelSize = new PixelSize((int)bounds.Width, (int)bounds.Height);
                var renderTarget = new RenderTargetBitmap(pixelSize, dpi);

                renderTarget.Render(visual);
                using (var fileStream = File.OpenWrite(@"OUTPUT.png"))
                {
                    renderTarget.Save(fileStream);
                }

Here is some example xaml which I tested and works in the live editor by avalonia which just displays a blue square.

<img width="286" alt="2024-03-15 16_55_01-ava2 xaml - Notepad" src="https://github.com/AvaloniaUI/Avalonia/assets/69156068/4275fa73-5c53-4bdd-9de6-f6601fcc26e4">

I however receive an error message:
XamlX.XamlParseException: Internal compiler error while transforming node XamlX.Ast.XamlAstNamePropertyReference:
XamlX.XamlTypeSystemException: Unable to resolve type Avalonia.Data.RelativeSource

**Comment by maxkatz6** [comment]
Try latest nightly or beta builds.

**Reply by denver116** [reply]
Hi thanks for responding, just a bit unsure, is that something I can do from the package manager in visual studio, for more context this is just a console app. Also here are the versions I am using,     
PackageReference Include="Avalonia" Version="11.0.10" 
PackageReference Include="Avalonia.Controls.DataGrid" Version="11.0.10" 
PackageReference Include="Avalonia.Desktop" Version="11.0.10" 
PackageReference Include="Avalonia.Markup.Xaml.Loader" Version="11.0.10"

**Reply by maxkatz6** [reply]
Try 11.1.0-beta1

**Reply by denver116** [reply]
Hi, got that installed and still get the same error.

**Reply by maxkatz6** [reply]
Add `[DynamicDependency(typeof(Avalonia.Data.RelativeSource))]` attribute on executing method then.

**Reply by denver116** [reply]
Did you mean like so [DynamicDependency("Avalonia.Data.RelativeSource, Avalonia.Base")]? Still receive an error and if I tried the one you pasted I get "cannot convert from System.Type to string.

**Reply by denver116** [reply]
Not sure the reasoning but adding this to my xaml worked, is there another way to avoid having to do this in any xaml? 
<img width="428" alt="a" src="https://github.com/AvaloniaUI/Avalonia/assets/69156068/f2c07db1-1101-403d-8177-6ab620d1ee90">
I also had to change visual to a Control() and cast it to (Control).

**Reply by maxkatz6** [reply, marked as answer]
> is there another way to avoid having to do this in any xaml?

You can root this type from being removed by the compiler in C# too:
`GC.KeepAlive(typeof(RelativeSource));`

**Reply by denver116** [reply]
Thanks that works great.
